crates/bevy_state/src/app.rs

```diff
@@ -87,6 +91,16 @@ impl AppExtStates for SubApp {
                 exited: None,
                 entered: Some(state),
             });
+        } else {
+            // Overwrite previous state and setup event
+            self.insert_resource::<State<S>>(State::new(state.clone()));
```

Comment:
Noting that there's some duplication between the two branches of the `if-else`. Pulling them out of the `if-else` would probably _hurt_ readability, so I won't suggest that :)

Reply:
There is a lot of overlap between all the state methods, we can do a cleanup PR later trying to factorize out the common parts